test/cql-pytest/test_permissions.py

```diff
@@ -39,3 +40,66 @@ def test_grant_applicable_data_and_role_permissions(cql, test_keyspace, cassandr
             for permission in all_permissions.difference(set(permissions)):
                 with pytest.raises((InvalidRequest, SyntaxException), match="support.*permissions"):
                     cql.execute(f"GRANT {permission} ON {resource} TO {user}")
+
+
+def eventually_authorized(fun, timeout_s=10):
+    for i in range(timeout_s * 10):
+        try:
+            return fun()
+        except Unauthorized as e:
+            time.sleep(0.1)
+    return fun()
+
+def eventually_unauthorized(fun, timeout_s=10):
+    for i in range(timeout_s * 10):
+        try:
+            fun()
+            time.sleep(0.1)
+        except Unauthorized as e:
+            return
+    try:
+        fun()
+        pytest.fail(f"Function {fun} was not refused as unauthorized")
+    except Unauthorized as e:
+        return
+
+# Test that data permissions can be granted and revoked, and that they're effective
+def test_udf_grant_revoke_data_permissions(cql, test_keyspace):
```

Comment:
The name of this function seems to be wrong - since it's not related to udf or "data permissions". You can have one big test (or multiple smaller tests) for all the various types of permissions, and have the UDF test seperately (if I understand correctly, the UDF part will still fail because this feature isn't completely implemented?)

Reply:
oh, indeed, it has no connection to udf whatsoever; as for "data", it is related - in Scylla and Cassandra nomenclature, ALL KEYSPACES, KEYSPACE ks and TABLE t are all "data resources", as opposed to "role resources" and "function resources".

Reply:
Weird. In my mind "data" is the rows, the table itself is "metadata". But ok :-)